models/buffer_disk.go

```diff
@@ -167,33 +184,80 @@ func (b *DiskBuffer) Batch(batchSize int) []telegraf.Metric {
 		}
 
 		metrics = append(metrics, m)
+		offsets = append(offsets, offset)
 		b.batchSize++
 		batchSize--
 	}
-	return metrics
+	return &Transaction{Batch: metrics, valid: true, state: offsets}
 }
 
-func (b *DiskBuffer) Accept(batch []telegraf.Metric) {
+func (b *DiskBuffer) EndTransaction(tx *Transaction) {
+	if len(tx.Batch) == 0 {
+		return
+	}
+
+	// Ignore invalid transactions and make sure they can only be finished once
+	if !tx.valid {
+		return
+	}
+	tx.valid = false
+
+	// Get the metric offsets from the transaction
+	offsets := tx.state.([]int)
+
 	b.Lock()
 	defer b.Unlock()
 
-	if b.batchSize == 0 || len(batch) == 0 {
-		// nothing to accept
+	// Mark metrics which should be removed in the internal mask
+	remove := make([]int, 0, len(tx.Batch)-len(tx.Accept)-len(tx.Reject))
+	for _, idx := range tx.Accept {
+		b.metricWritten(tx.Batch[idx])
+		remove = append(remove, offsets[idx])
+	}
+	for _, idx := range tx.Reject {
+		b.metricRejected(tx.Batch[idx])
+		remove = append(remove, offsets[idx])
+	}
+	b.mask = append(b.mask, remove...)
+	sort.Ints(b.mask)
```

Comment:
Is this sort necessary? I worry a bit about performance in something this low-level

Reply:
Good point. We need the mask to be ordered for the code in line 234ff to work otherwise we do need a `O(N^2)` complexity below to check for consecutive indices in this array when determining the `removeIdx` below. Do you know of a better way without sorting?

Reply:
Hmm that is a bit complicated, sort here may be the best approach for the combination of reasonable performance and also readability/maintainability then